### PR TITLE
Add FalconBackbone

### PR DESCRIPTION
**Pull request by Ruqyai**
Step 2: PR 1 - Add FalconBackbone
Please review the [Colab](https://colab.research.google.com/drive/1psl7CuukRE-NgMh9S6bOpYaSwdfP9DML?usp=sharing)

### CODE REVIEW

**Comment by Ruqyai** [comment]
Hi all
question about:
[class RWPreTrainedModel(PreTrainedModel):](https://huggingface.co/tiiuae/falcon-7b/blob/main/modelling_RW.py#L419-L490)  
Do you think I need to add all these details to
[keras_nlp/models/falcon/falcon_backbone.py](https://github.com/keras-team/keras-nlp/pull/1081/files#diff-dff33c35dfde4f894394e4ddba94a51618ba03f8093ebff34a4e02e720dd33ef)  ?

**Comment by Ruqyai** [comment]
@mattdangerw 
Is there a solution?
If not, I'll go to the cloud ,I have credit.

image

**Comment by Ruqyai** [comment]
I added the link of [Colab](https://colab.research.google.com/drive/1psl7CuukRE-NgMh9S6bOpYaSwdfP9DML?usp=sharing) in the PR . 
It is still draft.

**Comment by Ruqyai** [comment]
@mattdangerw  @abheesht17

though I change the values of `hidden_size` and `num_heads` many time, I get

ValueError: `hidden_size` must be divisible by num_heads (got `hidden_size`: ( )and `num_heads`: ()).

Do you catch why is an error? [Colab](https://colab.research.google.com/drive/1psl7CuukRE-NgMh9S6bOpYaSwdfP9DML?usp=sharing)

**Comment by Ruqyai** [comment]
image

https://colab.research.google.com/drive/1psl7CuukRE-NgMh9S6bOpYaSwdfP9DML?usp=sharing

this is will help you to review and also I added on the colab links to each part in hugging face to compare the coding
e.g.
"""Falcon Decoder Layer
https://huggingface.co/tiiuae/falcon-7b/blob/main/modelling_RW.py#L350-L416
"""

**Comment by shivance** [comment]
> I added the link of [Colab](https://colab.research.google.com/drive/1psl7CuukRE-NgMh9S6bOpYaSwdfP9DML?usp=sharing) in the PR . It is still draft.

@Ruqyai  While this works, I would recommend you to make your code compilant with our design, before opening a PR.
Please refer #1056 
Secondly, we'll be moving in RotaryEmbedding to `layers/` pretty soon, so you can remove your custom layer #1092

File: keras_nlp/models/falcon/falcon_mlp.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+"""Falcon MLP"""
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras.layers import Dense
+from tensorflow.keras.activations import gelu
+from keras_nlp.api_export import keras_nlp_export
+
+
+@keras_nlp_export("keras_nlp.layers.FalconMLP")
+class FalconMLP(keras.layers.Layer):
```
Comment (shivance) [inline keras_nlp/models/falcon/falcon_mlp.py:24]: We can move this to decoder layer

File: keras_nlp/models/falcon/falcon_rotary_positional_embedding.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+"""Position embedding implementation based on `keras.layers.Layer`."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_nlp.api_export import keras_nlp_export
+
+
+@keras_nlp_export("keras_nlp.layers.RotaryPositionalEmbedding")
+class FalconRotaryPositionalEmbedding(keras.layers.Layer):
```
Comment (shivance) [inline keras_nlp/models/falcon/falcon_rotary_positional_embedding.py:24]: #1092 is on the way !

File: keras_nlp/models/falcon/falcon_attention.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+                f" {self.num_heads})."
+            )
+
+        self.maybe_rotary = FalconRotaryPositionalEmbedding(config.head_dim) if config.rotary else lambda q, k: (q, k)
+
+        # Layer-wise attention scaling
+        self.inv_norm_factor = 1.0 / math.sqrt(self.head_dim)
+        self.beta = self.inv_norm_factor
+
+        self.query_key_value = Dense(
```
Comment (shivance) [inline keras_nlp/models/falcon/falcon_attention.py:86]: We can probably split this layer into three seperate EinsumDense layers. Can you verify?

File: keras_nlp/models/falcon/falcon_attention.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        query_layer = tf.transpose(query_layer, (1, 2, 0, 3))
+        query_layer = tf.reshape(query_layer, (q_length, batch_size * self.num_heads, self.head_dim))
+        key_layer = tf.transpose(key_layer, (1, 2, 0, 3))
+        key_layer = tf.reshape(key_layer, (q_length, batch_size * self.num_kv, self.head_dim))
+        value_layer = tf.transpose(value_layer, (1, 2, 0, 3))
+        value_layer = tf.reshape(value_layer, (q_length, batch_size * self.num_kv, self.head_dim))
+
+        query_layer, key_layer = self.maybe_rotary(query_layer, key_layer)
+
+        if layer_past is not None:
```
Comment (shivance) [inline keras_nlp/models/falcon/falcon_attention.py:163]: At this stage we probably won't need this

File: keras_nlp/models/falcon/falcon_backbone.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    def __init__(
+        self,
+        vocab_size=250880,
+        hidden_size=64,
+        n_layer=2,  # num_hidden_layers
+        n_head=8,  # num_attention_heads
+        layer_norm_epsilon=1e-5,
+        initializer_range=0.02,
+        use_cache=True,
+        bos_token_id=1,
```
Comment (shivance) [inline keras_nlp/models/falcon/falcon_backbone.py:105]: Again, this is not yet compilant with our design.

**Review by shivance — CHANGES_REQUESTED** [review]

**Comment by mattdangerw** [comment]
This would probably need some reworks from the source code to match the model style we use in this repo.

- Always take in named parameters to init, not `config` objects.
- Make sure argument names are self consistent, and consistent across the library. `num_layers` not `n_layers`, etc. Our [style guide](https://github.com/keras-team/keras-nlp/blob/master/STYLE_GUIDE.md) the [keras API guide](https://github.com/keras-team/governance/blob/master/keras_api_design_guidelines.md) cover some of this.
- Try to align the flow of our calls as much with `TransformerDecoder` and `MultiHeadAttention` as possible. Totally acceptable to have a separate transformer block, but we should try to match the general flow, and arguments as much as possible.

We should also make sure to try to support a fairly minimal set of configuration options. E.g. I see options for RoPE and ALiBi here, but if our checkpoints only need one of these, we should only support what we need. Better to be minimal than exhaustive here, especially for review!

**Comment by Ruqyai** [comment]
> This would probably need some reworks from the source code to match the model style we use in this repo.
> 
> * Always take in named parameters to init, not `config` objects.
> * Make sure argument names are self consistent, and consistent across the library. `num_layers` not `n_layers`, etc. Our [style guide](https://github.com/keras-team/keras-nlp/blob/master/STYLE_GUIDE.md) the [keras API guide](https://github.com/keras-team/governance/blob/master/keras_api_design_guidelines.md) cover some of this.
> * Try to align the flow of our calls as much with `TransformerDecoder` and `MultiHeadAttention` as possible. Totally acceptable to have a separate transformer block, but we should try to match the general flow, and arguments as much as possible.
> 
> We should also make sure to try to support a fairly minimal set of configuration options. E.g. I see options for RoPE and ALiBi here, but if our checkpoints only need one of these, we should only support what we need. Better to be minimal than exhaustive here, especially for review!

Thanks, I fixed the __init__ parameters, but I still have an error in the attention module. It's about the dimensions and shapes of the tensors involved in the reshape operation. 
@shivance @mattdangerw can you help in this point?

**Comment by Ruqyai** [comment]
Good News
These I watched these links and fixed a lot of errors :)
https://youtu.be/yJgSGE2-L_o
https://youtu.be/8IMei5V-2ok
https://youtu.be/GVShIIh3_yE

File: keras_nlp/models/falcon/falcon_decoder.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+class FalconDecoderLayer(tf.keras.Model):
+    """FalconDecoder used in Falcon models for decoding.
+
+    This layer implements the decoder layer used in Falcon models. It consists of
+    self-attention mechanism followed by a feed-forward neural network (MLP).
+
+    Args:
+        config: Configuration object containing the hyperparameters.
+
+    Inputs:
```
Comment (shivance) [inline keras_nlp/models/falcon/falcon_decoder.py:35]: Please add examples or we could just drop it for now.

File: keras_nlp/models/falcon/falcon_decoder.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+
+@keras_nlp_export("keras_nlp.layers.FalconDecoderLayer")
+class FalconDecoderLayer(tf.keras.Model):
+    """FalconDecoder used in Falcon models for decoding.
+
+    This layer implements the decoder layer used in Falcon models. It consists of
+    self-attention mechanism followed by a feed-forward neural network (MLP).
+
+    Args:
```
Comment (shivance) [inline keras_nlp/models/falcon/falcon_decoder.py:32]: We don't use config.

File: keras_nlp/models/falcon/falcon_decoder.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        )
+
+        self.dense_h_to_4h = Dense(4 * hidden_dim, use_bias=False)
+        self.act = gelu
+        self.dense_4h_to_h = Dense(hidden_dim, use_bias=False)
+        self.hidden_dropout = Dropout(self.dropout)
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
```
Comment (shivance) [inline keras_nlp/models/falcon/falcon_decoder.py:121]: Please add rest of attributes here.

File: keras_nlp/models/falcon/falcon_mlp.py (hunk 1 above)
Comment (shivance) [reply keras_nlp/models/falcon/falcon_mlp.py:24]: @Ruqyai update on this?
I recommend you to go through other transformers. We don't create a seperate MLP class.

File: keras_nlp/models/falcon/falcon_rotary_positional_embedding.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+
+    This layer computes rotary positional embeddings that can be used in Transformer models
+    to incorporate positional information into the input sequences.
+
+    Args:
+        head_dim (int): The dimensionality of the positional embeddings.
+        base (int): The base value used in the exponential calculation. Defaults to 10000.
+
+    Input Shape:
```
Comment (shivance) [inline keras_nlp/models/falcon/falcon_rotary_positional_embedding.py:35]: The docstring format is incorrect.

File: keras_nlp/models/falcon/falcon_rotary_positional_embedding_test.py
Comment (shivance) [inline keras_nlp/models/falcon/falcon_rotary_positional_embedding_test.py:1]: Could you delete this empty file ?

File: keras_nlp/models/falcon/falcon_attention.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+            scores += mask * tf.float32.min
+        
+        attention_weights = tf.nn.softmax(scores, axis=-1)
+        attention_weights = tf.nn.dropout(attention_weights, rate=dropout_rate)
+        
+        output = tf.einsum('...qk,...kd->...qd', attention_weights, value)
+        
+        return output, attention_weights
+    
+    def call(self, hidden_states, attention_mask=None, alibi=None, layer_past=None, head_mask=None,
```
Comment (shivance) [inline keras_nlp/models/falcon/falcon_attention.py:122]: Please run `black` to format code.

**Review by shivance — CHANGES_REQUESTED** [review]

**Comment by shivance** [comment]
The rule of thumb while contributing a model is to open PR for backbone first, once that gets in, one should proceed for preprocessors, tasks etc.

**Comment by shivance** [comment]
Additionally, we should migrate this PR to use Keras Core instead of tf.keras

**Pull request closed without merging**